codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java

```diff
@@ -431,7 +485,50 @@ public ChannelFuture close(Channel channel, CloseWebSocketFrame frame, ChannelPr
         if (channel == null) {
             throw new NullPointerException("channel");
         }
-        return channel.writeAndFlush(frame, promise);
+        channel.writeAndFlush(frame, promise);
+        applyForceCloseTimeout(channel, promise);
+        return promise;
+    }
+
+    private void applyForceCloseTimeout(final Channel channel, ChannelFuture flushFuture) {
+        final long forceCloseTimeoutMillis = this.forceCloseTimeoutMillis;
+        if (forceCloseTimeoutMillis <= 0 || !channel.isActive() || forceCloseInit.get()) {
+            return;
+        }
+
+        flushFuture.addListener(new ChannelFutureListener() {
+            @Override
+            public void operationComplete(ChannelFuture future) throws Exception {
+                if (!future.isSuccess() || !channel.isActive()) {
+                    // If flush operation failed, there is no reason to expect
+                    // a server to receive CloseFrame. Thus this should be handled
+                    // by the application separately.
+                    return;
+                }
+
+                if (!forceCloseInit.compareAndSet(false, true)) {
+                    // Close might be called twice from different threads.
+                    return;
+                }
```

Comment:
nit: more of a coding style preference, but I'd rather have one single boolean operation wrapping the method's body: 
`if (future.isSuccess() && channel.isActive() && forceCloseInit.compareAndSet(false, true))`